syntax = "proto3";

package raft;

service Raft {
    // RequestVote RPC
    rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse);
    // AppendEntries RPC
    rpc AppendEntries (AppendEntriesRequest) returns (AppendEntriesResponse);
}

// RequestVote message
message RequestVoteRequest {
    int32 term = 1;
    string candidate_id = 2;
    int32 last_log_index = 3;
    int32 last_log_term = 4;
}

// RequestVote response
message RequestVoteResponse {
    string nodeId = 1;
    int32 term = 2;
    bool vote_granted = 3;
}

// AppendEntries message
message AppendEntriesRequest {
    int32 term = 1;
    string leader_id = 2;
    int32 prev_log_index = 3;
    int32 prev_log_term = 4;
    repeated LogEntry entries = 5;
    int32 leader_commit = 6;
}

// AppendEntries response
message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

// LogEntry message
message LogEntry {
    int32 term = 1;
    string command = 2;
}
